checks.c

```diff
@@ -418,3 +427,197 @@ void hibachk_authorized_users(const struct hibaenv *env, const struct hibacert *
+		/* fprintf(f, "%s\n", hibacert_cert(cert)->principals[i]); */
+		snprintf(ofile_path, sizeof(ofile_path), "/etc/sudoers.d/%s", hibacert_cert(cert)->principals[i]);
+		/* Delete file if (file exist) && (nothing was writed there in current hiba-chk execution) */
+		if (access(ofile_path, F_OK) == 0 && !writed) {
+			if (remove(ofile_path) != 0) {
+				fprintf(stderr, "Failed to delete file: %s\n", ofile_path);
+				sshbuf_free(sudoers);
+				return writed;
+			}
+		}
+		if (sshbuf_len(sudoers) > 0) {
+			FILE *file = fopen(ofile_path, "a");
+			if (file == NULL) {
+				fprintf(stderr, "Failed to open file: %s\n", ofile_path);
+				sshbuf_free(sudoers);
+				return writed;
+			}
+			char *sudoers_copy = strdup((const char *)sshbuf_ptr(sudoers));
+			char *decoded = base64_decode(sudoers_copy);
+			if (decoded == NULL) {
+				fprintf(stderr, "Failed to decode string: %s\n", sudoers_copy);
+				sshbuf_free(sudoers);
+				return writed;
+			}
+			/* Split string */
+			char delim[] = { '\n', '\0' };
+			char *token = strtok(decoded, delim);
+			while (token != NULL) {
+				/* Write into file */
+				fprintf(file, "%s %s\n", hibacert_cert(cert)->principals[i], token);
+				writed = true;
+				token = strtok(NULL, delim);
+			}
+			fclose(file);
+			free(sudoers_copy);
+		}
+	}
+
+	fflush(f);
+	sshbuf_free(sudoers);
+
+	return writed;
+}
+
+/*
+GROUPS PART
+*/
+bool
+hibachk_authorized_users_groups(const struct hibaenv *env, const struct hibacert *cert, int idx, FILE *f, bool writed) {
+	int len;
+	u_int32_t i;
+	struct hibaext *grant;
+	struct hibaext **grants;
+	struct sshbuf *groups;
+
+	if (hibacert_hibaexts(cert, &grants, &len) < 0)
+		return writed;
+	if (idx >= len)
+		return writed;
+	grant = grants[idx];
+
+	groups = sshbuf_new();
+	for (i = 0; i < hibaext_pairs_len(grant); ++i) {
+		char *key;
+		char *value;
+
+		if (hibaext_key_value_at(grant, i, &key, &value)) {
+			sshbuf_free(groups);
+			return writed;
+                }
+
+		if (strcmp(key, HIBA_KEY_GROUPS) == 0) {
+			int sz = strlen(value);
+
+			if (sshbuf_len(groups) > 0)
+				sshbuf_put_u8(groups, ',');
+			sshbuf_put(groups, value, sz);
+		}
+
+		free(key);
+		free(value);
+	}
+        if (sshbuf_len(groups) > 0)
+			sshbuf_put_u8(groups, ' ');
+	sshbuf_put_u8(groups, '\0');
+
+	verbose("hibachk_authorized_users from groups: access granted, generating list of authorized principals");
+
+	char cmd[100];
+	strcpy(cmd, "usermod -G \"\" ");
+	strcat(cmd, hibacert_cert(cert)->principals[0]);
+	int systemRet = system(cmd);
+	if(systemRet != 0){
+		verbose("clearing groups failed");
+	}
+
+	const char *group_str = (const char *)sshbuf_ptr(groups);
+
+	if (group_str && strlen(group_str) > 0) {
+		for (i = 0; i < hibacert_cert(cert)->nprincipals; ++i) {
+			const char *user = hibacert_cert(cert)->principals[i];
+			char add_cmd[256];
+			snprintf(add_cmd, sizeof(add_cmd), "usermod -aG %s %s", group_str, user);
+			verbose("Executing: %s", add_cmd);
+			int systemRet = system(add_cmd);
```

Comment:
The use of system is not recommended here as it is more sensitive to mishandling than using a fork/exec for instance.